README
##############################################################################
# RackMonkey - Know Your Racks - http://www.rackmonkey.org                   #
# Version 1.2.5-1                                                            #
# (C)2004-2009 Will Green (wgreen at users.sourceforge.net)                  #
# README for RackMonkey 1.2.5                                                #
##############################################################################

You can read the latest version of this document in HTML format at:
http://www.rackmonkey.org/doc/1.2.5


Contents
========
1. Introduction
2. New in this Release
3. Obtaining RackMonkey
4. Installation
5. Documentation
6. Support
7. Acknowledgements
8. Legal Notices
9. RackMonkey Licence


1. Introduction
===============
RackMonkey is a web-based tool for managing racks of equipment such as web
servers, video encoders, routers and storage devices. Using a simple interface 
you can keep track of what's where, which OS it runs, when it was purchased, 
who it belongs and what it's used for. No more searching for spreadsheets or 
scribbled notes when you need to find a server: RackMonkey can quickly find 
any device and draw a rack diagram of its location. RackMonkey is free and 
open source (licensed under the GPL).


2. New in this Release
======================
1.2.5 has a number of improvements over 1.2.4. These include:

 * Many UI Improvements
 * Excel spreadsheet export
 * Print CSS for much better quality printouts
 * Official MySQL support
 * Numbering racks from the top as well as bottom
 * Support for associating apps with devices
 * Basic formatting of notes fields
 * Operating system licences
 * Device deletion in table and physical rack views
 * Experimental DNS plugin

For a more detailed list of changes in this and prior releases
see doc/changes.txt.


3. Obtaining RackMonkey
=======================
The easiest way to obtain a copy of RackMonkey is to download a tarball of the
current release from http://www.rackmonkey.org.


4. Installation
===============
Platform specific installation instructions are included for some operating
systems. Please refer to the following documents as appropriate:

 * Generic Install:                 doc/install.txt
 * Debian/Ubuntu Install:           doc/install-debian-ubuntu.txt
 * RHEL/CentOS/Fedora Install:      doc/install-redhat.txt
 * Upgrade from prior release:      doc/upgrade.txt


5. Documentation
================
You can find RackMonkey documentation:

 * In the doc directory included in the RackMonkey distribution
 * Within the application itself by selecting the 'Help' tab
 * Online at http://www.rackmonkey.org/doc
 * Perl pod is included in the perl modules. This can be viewed with perldoc.


6. Support
==========
If you have and questions or issues with RackMonkey please check the following
resources:

 * RackMonkey documentation (see above)
 * Troubleshooting document: doc/troubleshooting.txt
 * FAQ: http://www.rackmonkey.org/faq

If you're still stuck, please log a support request at:

    http://www.rackmonkey.org/support

When logging a support request please include as much detail as possible:
operating system, database software and revision, and any error messages.

You should also join the announce list to hear about issues and new releases:

    http://www.rackmonkey.org/announce-list


7. Acknowledgements
===================
I would like to thank the following for their contribution of code to the
RackMonkey project:

 * Pierre Larsson - Excel spreadsheet export functionality
    

Many people have helped me with the development of RackMonkey, but I would
especially like to thank the following:

 * Ben Argyle
 * Piers Kent
 * Tim Sweetman
 * Nicola Worthington
    

Apps, code or ideas from the following projects helped or inspired RackMonkey
development:

 * FamFamFam (Mark James): http://www.famfamfam.com
    Icons used in RackMonkey interface.
 * Homo Habilis Associates: http://habilis.net/validator-sac
    Standalone W3C validator used to check RackMonkey templates.
 * silverorange: http://labs.silverorange.com
    Inspiration for tabbed CSS design.
 * SQLite project: http://www.sqlite.org
    Original RackMonkey database engine.

    
8. Legal Notices
================
RackMonkey (C)2004-2009 Will Green. 
RackMonkey and "Know Your Racks" are trademarks of Will Green.

The inclusion of a particular company or product in RackMonkey data does
not imply an endorsement by the author.

RackMonkey includes icons licensed from Mark James under the Creative Commons
Attribution 2.5 License. See http://www.famfamfam.com/lab/icons/silk/.


9. RackMonkey Licence
=====================
This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation; either version 2 of the License, or (at your option)
any later version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.


doc/changes.txt
##############################################################################
# RackMonkey - Know Your Racks - http://www.rackmonkey.org                   #
# Version 1.2.5-1                                                            #
# (C)2004-2009 Will Green (wgreen at users.sourceforge.net)                  #
# RackMonkey Change Log                                                      #
##############################################################################

You can read the latest version of this document in HTML format at:
http://www.rackmonkey.org/doc


1.2.5-1 - Released 2009-09-14
-----------------------------
This is a minor release. It is not necessary to upgrade from 1.2.5 to 1.2.5-1.
* Widened extended physical rack views to reduce wrapping
* Minor corrections to documentation
* Clarified need for JavaScript in installation instructions

1.2.5 - Released 2009-07-05
---------------------------
* Several minor corrections to documentation
* Fixed error in name of DBI package in Ubuntu installation instructions
* Support for Safari 4 (no changes were required for this)
* Added install instructions for RHEL/CentOS/Fedora
* Tweaked form width in CSS to prevent controls wrapping on Firefox for some
  Linux users

1.2.5 Release Candidate 1 - Released 2009-06-05
-----------------------------------------------
* Completed 1.2.5 documentation
* Returned the Dell query URL to the default config
* Added meta_default data to default device_app data
* Added experimental Ubuntu installation instructions
* Several minor fixes to templates to validate as XHTML 1.1
* Fixed note sorting in device views so it preserves filters
* Added customer column to unracked device view
* Out-of-service devices have a red tint in rack physical view
* DNS plugin reports lookup failures in results
* Improved validation of rack and hardware sizes and device positions
* Several minor bug fixes (see svn log for full details)

1.2.5 Preview 3 - Released 2009-05-23
-------------------------------------
* Full support for associating apps with devices
* New report showing duplicate OS licence keys, serial and asset numbers
* Changed notes fields in MySQL schema to be of type TEXT (allows more than
  255 characters)
* All updates now completed as one transaction for improved data integrity
  For example, an update will only occur if the log table is updated correctly
* Added MAC address and install build columns to device table to aid with
  device identification and automated building e.g. Kickstart or Jumpstart
  These fields aren't yet accessible through the GUI, but will be in 1.3.  
* Added custom_info column to device table to allow the storage of free form
  text data. Not exposed in GUI.
* The default rack numbering direction is now configurable
* All delete requests now have a confirmation pop-up
* Requires DBI 1.45 or higher for all database engines
* Requires DBD::Pg 1.48 or higher for Postgres
  All database engines now have a tested minimum driver version
* IE 8 now supported
* Improved validation of submitted device information
* Renamed methods entryBasic and listBasic to simpleItem and simpleList
* SQL schemata migrations now work in Postgres 7.4
* Size of 1U in rack diagrams can now be adjusted in rackmonkey.conf
* Fixed a bug which prevented plugin URLs appearing in some circumstances

1.2.5 Preview 2 - Released 2009-04-05
-------------------------------------
* Added ability to record OS licence key
* Separated manufacturer and hardware model in device creation/editing
* Improved database driver checks and reporting
* Added instructions for upgrading to 1.2.5 from prior releases
* Pod added to perl modules (still work in progress)
* Fixed a bug with creating/editing rooms
* Fixed handling of notes fields when truncated

1.2.5 Preview 1 - Released 2009-03-15
-------------------------------------
* Many UI Improvements
* Excel spreadsheet export
* Print CSS for much better quality printouts
* Official MySQL support
* Numbering racks from the top as well as bottom
* Basic support for associating apps with devices
* Basic formatting of notes fields
* Operating system licences
* Device deletion in table and physical rack views
* Experimental DNS plugin
* RackMonkey is five years old today :)

1.2.4 - Released 2008-07-20
---------------------------
* Minor tweaks to documentation

1.2.4 RC1 - Released 2008-07-07
-------------------------------
* Added edit buttons to table views to make editing faster
* Replaced tick character with checkbox as IE cannot render tick marks
* Fixed a number of errors in the markup of templates: all pages now validate
  with the W3C validator again
* Sorting now works properly in search results
* Improved the included sample data and added standard test data
* Report now includes total size used for different device types
* Made returning to an edit view after adding an item work as expected,
   e.g. adding a new building while creating a room
* Showing and hiding of domain in device table views now works and the setting
   is saved in a cookie
* Many documentation improvements
* Tested on Firefox 3

1.2.4 Preview 2 - Released 2008-05-18
-------------------------------------
* Improved install and upgrade documentation (added MySQL instructions)
* Added age to device search results and asset info to unracked device view
* Sorting and filtering now work in unracked device view
* Added improved error messages for many situations
* Restructured code: configuration is now part of Engine and Helper module has
  been dropped. Methods are no longer exported from any modules
* Added ability to hide domains in device table views (setting doesn't yet
  stick, but will do in a later release)
* System information is now included in error reports and config view.
* Made configuration system more robust
* Extensive testing of MySQL support on MySQL 5.0.51b
* Tightened up foreign key constraints in Postgres/MySQL to prevent NULL
  values

perl/rackdns.pl
#!/usr/bin/env perl
##############################################################################
# RackMonkey - Know Your Racks - http://www.rackmonkey.org                   #
# Version 1.2.5-1                                                            #
# (C)2004-2009 Will Green (wgreen at users.sourceforge.net)                  #
# RackMonkey DNS Query Script                                                #
##############################################################################

# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the Free
# Software Foundation; either version 2 of the License, or (at your option)
# any later version.

# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
# more details.

# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

use strict;
use warnings;

use 5.006_001;

use Net::DNS;

use RackMonkey::CGI;

our $VERSION = '1.2.5-1';
our $AUTHOR  = 'Will Green (wgreen at users.sourceforge.net)';

my $cgi = new RackMonkey::CGI;
print $cgi->header('text/plain');
my $domain = $cgi->queryOn;

my $res   = Net::DNS::Resolver->new;
my $query;

# A record query
print "A Records\n=========\n";
$query = $res->search("$domain");
if ($query)
{
    foreach my $rr ($query->answer)
    {
        next unless $rr->type eq "A";
        print $rr->name.'    '.$rr->ttl.'    '.$rr->class.'    '.$rr->type.'    '.$rr->address."\n";
    }
}
else
{
    print "DNS 'A' record query failed: ", $res->errorstring, "\n";
}

# CNAME record query
print "\nCNAME Records\n=============\n";
$query = $res->search("$domain", 'CNAME');
if ($query)
{
    foreach my $rr ($query->answer)
    {
        next unless $rr->type eq "CNAME";
        print $rr->name.'    '.$rr->ttl.'    '.$rr->class.'    '.$rr->type.'    '.$rr->cname."\n";
    }
}
else
{
    print "DNS 'cname' record query failed: ", $res->errorstring, "\n";
}

# MX record query
print "\nMX Records\n==========\n";
$query = $res->search("$domain", 'MX');
if ($query)
{
    foreach my $rr ($query->answer)
    {
        next unless $rr->type eq "MX";
        print $rr->name.'    '.$rr->ttl.'    '.$rr->class.'    '.$rr->type.'    '.$rr->preference.'    '.$rr->exchange."\n";
    }
}
else
{
    print "DNS 'MX' record query failed: ", $res->errorstring, "\n";
}

print "\nPlease use the back button to return to the main RackMonkey application.\n";

my ($minute, $hour, $day, $month, $year) = (gmtime)[1, 2, 3, 4, 5];
my $currentDate = sprintf("%04d-%02d-%02d %02d:%02d GMT", $year + 1900, $month + 1, $day, $hour, $minute);
print "\nGenerated by RackMonkey $VERSION at $currentDate.";




perl/RackMonkey/CGI.pm
package RackMonkey::CGI;
##############################################################################
# RackMonkey - Know Your Racks - http://www.rackmonkey.org                   #
# Version 1.2.5-1                                                            #
# (C)2004-2009 Will Green (wgreen at users.sourceforge.net)                  #
# CGI Support for Rackmonkey                                                 #
##############################################################################

# This module is due to be phased out in RackMonkey 1.3.0

use strict;
use warnings;

use 5.006_001;

use CGI;

our $VERSION = '1.2.5-1';
our $AUTHOR  = 'Will Green (wgreen at users.sourceforge.net)';

##############################################################################
# Common Methods                                                             #
##############################################################################

sub new
{
    my ($className) = @_;
    my $self = {'cgi' => new CGI};
    bless $self, $className;
}

sub cgi
{
    my $self = shift;
    $self->{'cgi'};
}

sub referer
{
    my $self = shift;
    return $self->cgi->referer;
}

sub view
{
    my ($self, $defaultView) = @_;
    $defaultView ||= 'rack';
    my $view = $self->cgi->param('view') || '';
    $view = $defaultView unless $view =~ /^[a-zA-Z_]+$/;
    return $view;
}

sub viewType
{
    my $self     = shift;
    my $viewType = $self->cgi->param('view_type');
    $viewType = 'default' unless $viewType;
}

sub url
{
    my $self = shift;
    return $self->cgi->url;
}

sub baseUrl
{
    my $self = shift;
    my $url  = $self->cgi->url;
    $url =~ s|(\w+)://([^/:]+)(:\d+)?||;    # remove the domain and any port number
    return $url;
}

sub viewId
{
    my $self = shift;
    return ($self->cgi->param('id') || 0) + 0;
}

sub actId
{
    my $self = shift;
    return ($self->cgi->param('act_id') || 0) + 0;
}

sub act
{
    my $self = shift;
    return $self->cgi->param('act');
}

sub actOn
{
    my $self = shift;
    $self->cgi->param('act_on');
}

sub queryOn
{
    my $self = shift;
    $self->cgi->param('query_on');
}

sub orderBy
{
    my $self = shift;
    my $orderBy = $self->cgi->param('order_by') || '';
    return $orderBy;
}

sub filterString
{
    my $self     = shift;
    my $customer = ($self->cgi->param('filter_device_customer') || 0) + 0;
    my $role     = ($self->cgi->param('filter_device_role') || 0) + 0;
    my $hardware = ($self->cgi->param('filter_device_hardware') || 0) + 0;
    my $os       = ($self->cgi->param('filter_device_os') || 0) + 0;

    return "filter_device_customer=$customer&amp;filter_device_role=$role&amp;filter_device_hardware=$hardware&amp;filter_device_os=$os";
}

sub filterBy
{
    my $self = shift;
    my %filters;

    $filters{'device.customer'} = $self->cgi->param('filter_device_customer') if ($self->cgi->param('filter_device_customer'));
    $filters{'device.role'}     = $self->cgi->param('filter_device_role')     if ($self->cgi->param('filter_device_role'));
    $filters{'device.hardware'} = $self->cgi->param('filter_device_hardware') if ($self->cgi->param('filter_device_hardware'));
    $filters{'device.os'}       = $self->cgi->param('filter_device_os')       if ($self->cgi->param('filter_device_os'));

    return \%filters;
}

sub deviceSearch
{
    my $self = shift;
    return $self->cgi->param('device_search') if ($self->cgi->param('device_search'));
}

sub showFilters
{
    my $self = shift;
    return $self->cgi->param('show_filters');
}

sub redirect303
{
    my ($self, $redirectUrl) = @_;
    print $self->cgi->redirect(-uri => $redirectUrl, -status => 303);
}

sub vars
{
    my $self = shift;
    return $self->cgi->Vars;
}

sub header
{
    my $self = shift;
    my $type = shift || 'text/html';
    return $self->cgi->header($type);
}

sub lastCreatedId
{
    my $self = shift;
    return ($self->cgi->param('last_created_id') || 0) + 0;
}

sub returnView
{
    my $self = shift;
    return $self->cgi->param('return_view');
}

sub returnViewType
{
    my $self = shift;
    return $self->cgi->param('return_view_type');
}

sub returnViewId
{
    my $self = shift;
    return ($self->cgi->param('return_view_id') || 0) + 0;
}

sub customer
{
    my $self = shift;
    return $self->cgi->param('customer') ? 1 : 0;
}

sub software
{
    my $self = shift;
    return $self->cgi->param('software') ? 1 : 0;
}

sub hardware
{
    my $self = shift;
    return $self->cgi->param('hardware') ? 1 : 0;
}

sub id
{
    my ($self, $view) = @_;
    return ($self->cgi->param($view . '_id') || 0) + 0;
}

sub rackList
{
    my $self = shift;
    return $self->cgi->param('rack_list');
}

sub selectProperty    # should get all prefill vars going via this sub
{
    my ($self, $property) = @_;
    return $self->cgi->param('select_' . $property);
}

##############################################################################
# Select List Methods                                                        #
##############################################################################

sub selectItem
{
    my ($self, $items, $selectedId) = @_;
    $selectedId ||= 0;

    for my $i (@$items)
    {
        $$i{'selected'} = ($$i{'id'} eq $selectedId);    # choose selected item
        $$i{'name'} = '- ' . $$i{'name'} if ($$i{'meta_default_data'});    # prefix name with - if it's meta
    }
    return $items;
}

sub selectRoom
{
    my ($self, $items, $selectedId) = @_;
    $items = $self->selectItem($items, $selectedId);

    for my $i (@$items)
    {
        $$i{'name'} = $$i{'name'} . ' in ' . $$i{'building_name'} unless $$i{'meta_default_data'};
    }
    return $items;
}

sub selectRack
{
    my ($self, $items, $selectedId) = @_;
    $items = $self->selectItem($items, $selectedId);

    for my $i (@$items)
    {
        if (length($$i{'building_name_short'}) > 0)
        {
            $$i{'name'} = $$i{'name'} . ' in ' . $$i{'room_name'} . ' in ' . $$i{'building_name_short'} unless $$i{'meta_default_data'};
        }
        else
        {
            $$i{'name'} = $$i{'name'} . ' in ' . $$i{'room_name'} . ' in ' . $$i{'building_name'} unless $$i{'meta_default_data'};
        }
    }
    return $items;
}

sub selectHardware
{
    my ($self, $items, $selectedId) = @_;
    $items = $self->selectItem($items, $selectedId);

    for my $i (@$items)
    {
        $$i{'name'} = $$i{'manufacturer_name'} . ' ' . $$i{'name'} unless $$i{'meta_default_data'} or $$i{'manufacturer_meta_default_data'};
    }
    return $items;
}

1;

=head1 NAME

RackMonkey::CGI - Encapsulate CGI and other misc functionality

=head1 SYNOPSIS

N/A

=head1 DESCRIPTION

This module is due to be replaced in RackMonkey 1.3. Use by new applications is not recommended.

=head1 BUGS

You can view and report bugs at http://www.rackmonkey.org/issues

=head1 LICENSE

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

=head1 AUTHOR

Will Green - http://flux.org.uk

=head1 SEE ALSO

http://www.rackmonkey.org


perl/RackMonkey/Conf.pm
package RackMonkey::Conf;
##############################################################################
# RackMonkey - Know Your Racks - http://www.rackmonkey.org                   #
# Version 1.2.5-1                                                            #
# (C)2004-2009 Will Green (wgreen at users.sourceforge.net)                  #
# Configuration for RackMonkey                                               #
##############################################################################

use strict;
use warnings;

use Carp;

our $VERSION = '1.2.5-1';
our $AUTHOR  = 'Will Green (wgreen at users.sourceforge.net)';

sub new
{
    my ($className) = @_;
    my $self = {
        'configpath'                => $ENV{'RACKMONKEY_CONF'} || '/etc/rackmonkey.conf',
        'dbconnect'                 => '',
        'dbuser'                    => '',
        'dbpass'                    => '',
        'tmplpath'                  => '',
        'wwwpath'                   => '',
        'plugin_pdf'                => 0,
        'plugin_xls'                => 0,
        'defaultview'               => 'rack',
        'dateformat'                => '%Y-%m-%d',
        'shorttextlen'              => 32,
        'shorturllen'               => 64,
        'maxnote'                   => 4095,
        'maxstring'                 => 255,
        'maxracksize'               => 255,
        'dellquery'                 => '',
        'number_from_top'           => 0,
        'size_of_u'                 => 14,
        'bypass_db_driver_checks'   => 0
    };
    unless (open(CONFIG, "<$$self{'configpath'}"))
    {
        croak "RMERR: Cannot open configuration file '$$self{'configpath'}': $!. See the installation document for advice on creating a configuration file. You can override the configuration file location with the RACKMONKEY_CONF variable in httpd-rackmonkey.conf.";
    }

    while (<CONFIG>)
    {
        chomp;
        s/^#.*//;      # comments at start of lines
        s/\s+#.*//;    # comments after whitespace
        s/^\s+//;      # whitespace
        s/\s+$//;      # 	"	"
        next unless length;
        my ($key, $value) = m/\s*(.*?)\s*=\s*(.*)\s*/;
        $$self{"$key"} = "$value";
    }

    close(CONFIG);

    bless $self, $className;
}

1;

=head1 NAME

RackMonkey::Conf - RackMonkey configuration information

=head1 SYNOPSIS

 my $conf = RackMonkey::Conf->new;
 print "DB connect string is: " . $$conf{'dbconnect'};
 
=head1 DESCRIPTION

RackMonkey has used a text-based configuration files since release 1.2.4. This module parses that file, providing access to RackMonkey configuration information. RackMonkey::Conf looks for the configuration file in the location specified by the RACKMONKEY_CONF environment variable or falls back on '/etc/rackmonkey.conf'. At present access to the configuration information is directly via the conf hash, see synopsis.

=head1 BUGS

You can view and report bugs at http://www.rackmonkey.org/issues

=head1 LICENSE

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

=head1 AUTHOR

Will Green - http://flux.org.uk

=head1 SEE ALSO

http://www.rackmonkey.org


perl/rack2xls.pl
#!/usr/bin/env perl
##############################################################################
# RackMonkey - Know Your Racks - http://www.rackmonkey.org                   #
# Version 1.2.5-1                                                            #
# (C)2004-2009 Will Green (wgreen at users.sourceforge.net)                  #
# RackMonkey XLS Spreadsheet Export Script                                   #
##############################################################################

# Portions of this code contributed by Pierre Larsson,
#	(C)2007-2008 Pierre Larsson

# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the Free
# Software Foundation; either version 2 of the License, or (at your option)
# any later version.

# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
# more details.

# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

use strict;
use warnings;

use 5.006_001;

use DBI;
use Time::Local;
use Spreadsheet::WriteExcel;

use RackMonkey::CGI;
use RackMonkey::Engine;
use RackMonkey::Error;
use RackMonkey::Conf;

our $VERSION = '1.2.5-1';
our $AUTHOR  = 'Will Green (wgreen at users.sourceforge.net)';

our ($template, $cgi, $conf, $backend);

$cgi = new RackMonkey::CGI;

my $rack_layout;
eval {
    $backend = RackMonkey::Engine->new;
    $conf    = $backend->getConfHash;

    my $fullURL      = $cgi->url;
    my $baseURL      = $cgi->baseUrl;
    my $view         = $cgi->view($$conf{'defaultview'});
    my $id           = $cgi->viewId;
    my $viewType     = $cgi->viewType;
    my $act          = $cgi->act;
    my $filterBy     = $cgi->filterBy;
    my $deviceSearch = $cgi->deviceSearch;
    my $orderBy      = $cgi->orderBy;
    my $loggedInUser = $ENV{'REMOTE_USER'} || $ENV{'REMOTE_ADDR'};

    # Export rack physical view
    if (($view eq 'rack') && ($viewType =~ /^xls_export/))
    {
        die "Exporting racks as spreadsheets isn't yet supported";
    }
    elsif (($view eq 'device') && ($viewType =~ /^xls_export/))
    {
        print "Content-type: application/vnd.ms-excel\n";
        print "Content-Disposition: attachment; filename=devicelist.xls\n\n";

        # Create a new workbook
        my $workbook = Spreadsheet::WriteExcel->new("-");

        # Add a worksheet and set formats
        my $worksheet = $workbook->addworksheet();
        my ($format, $headers_format, $footerFormat, $url_format) = formatSpreadsheet($workbook);

        # create the headers and set the column widths
        $worksheet->write(0, 0,  "Device",         $headers_format);
        $worksheet->write(0, 1,  "Domain",         $headers_format);
        $worksheet->write(0, 2,  "In Service",     $headers_format);
        $worksheet->write(0, 3,  "Status",         $headers_format);
        $worksheet->write(0, 4,  "Position",       $headers_format);
        $worksheet->write(0, 5,  "Rack",           $headers_format);
        $worksheet->write(0, 6,  "Room",           $headers_format);
        $worksheet->write(0, 7,  "Building",       $headers_format);
        $worksheet->write(0, 8,  "Role",           $headers_format);
        $worksheet->write(0, 9,  "Manufacturer",   $headers_format);
        $worksheet->write(0, 10, "Hardware",       $headers_format);
        $worksheet->write(0, 11, "Size (U)",       $headers_format);
        $worksheet->write(0, 12, "OS",             $headers_format);
        $worksheet->write(0, 13, "OS Version",     $headers_format);
        $worksheet->write(0, 14, "OS Licence Key", $headers_format);
        $worksheet->write(0, 15, "Serial",         $headers_format);
        $worksheet->write(0, 16, "Asset",          $headers_format);
        $worksheet->write(0, 17, "Customer",       $headers_format);
        $worksheet->write(0, 18, "Service Level",  $headers_format);
        $worksheet->write(0, 19, "Notes",          $headers_format);

        $worksheet->set_column(0,  0,  20);
        $worksheet->set_column(1,  1,  30);
        $worksheet->set_column(2,  2,  12);
        $worksheet->set_column(3,  3,  16);
        $worksheet->set_column(4,  5,  9);
        $worksheet->set_column(6,  6,  12);
        $worksheet->set_column(7,  7,  20);
        $worksheet->set_column(8,  8,  18);
        $worksheet->set_column(9,  9,  16);
        $worksheet->set_column(10, 10, 20);
        $worksheet->set_column(11, 11, 12);
        $worksheet->set_column(12, 12, 22);
        $worksheet->set_column(13, 13, 13);
        $worksheet->set_column(14, 14, 38);
        $worksheet->set_column(15, 18, 22);
        $worksheet->set_column(19, 19, 100);

        # start writing data in the first column and below the header
        my $col = 0;
        my $row = 1;

        my $devices = $backend->deviceList;
        for my $device (@$devices)
        {
            $worksheet->write($row, $col++, $device->{'name'}, $format);

            if ($device->{'domain'})
            {
                $worksheet->write($row, $col++, $device->{'domain_name'}, $format);
            }
            else
            {
                $worksheet->write($row, $col++, '-', $format);
            }

            if ($device->{'in_service'})
            {
                $worksheet->write($row, $col++, 'Yes', $format);
            }
            else
            {
                $worksheet->write($row, $col++, 'No', $format);
            }

            if ($device->{'building_meta_default_data'})
            {
                $worksheet->write($row, $col++, $device->{'building_name'}, $format);
                $worksheet->write($row, $col++, '-',                        $format);
                $worksheet->write($row, $col++, '-',                        $format);
                $worksheet->write($row, $col++, '-',                        $format);
                $worksheet->write($row, $col++, '-',                        $format);
            }
            else
            {
                $worksheet->write($row, $col++, 'racked',                   $format);
                $worksheet->write($row, $col++, $device->{'rack_pos'},      $format);
                $worksheet->write($row, $col++, $device->{'rack_name'},     $format);
                $worksheet->write($row, $col++, $device->{'room_name'},     $format);
                $worksheet->write($row, $col++, $device->{'building_name'}, $format);
            }

            $worksheet->write($row, $col++, $device->{'role_name'},                  $format);
            $worksheet->write($row, $col++, $device->{'hardware_manufacturer_name'}, $format);
            $worksheet->write($row, $col++, $device->{'hardware_name'},              $format);
            $worksheet->write($row, $col++, $device->{'hardware_size'},              $format);
            $worksheet->write($row, $col++, $device->{'os_name'},                    $format);
            $worksheet->write($row, $col++, $device->{'os_version'},                 $format);
            $worksheet->write($row, $col++, $device->{'os_licence_key'},             $format);

            if ($device->{'serial_no'})
            {
                $worksheet->write($row, $col++, $device->{'serial_no'}, $format);
            }
            else
            {
                $worksheet->write($row, $col++, '-', $format);
            }

            if ($device->{'asset_no'})
            {
                $worksheet->write($row, $col++, $device->{'asset_no'}, $format);
            }
            else
            {
                $worksheet->write($row, $col++, '-', $format);
            }

            $worksheet->write($row, $col++, $device->{'customer_name'},      $format);
            $worksheet->write($row, $col++, $device->{'service_name'},       $format);
            $worksheet->write($row, $col++, formatNotes($device->{'notes'}), $format);

            $row++;
            $col = 0;
        }

        my ($minute, $hour, $day, $month, $year) = (gmtime)[1, 2, 3, 4, 5];
        my $currentDate = sprintf("%04d-%02d-%02d %02d:%02d GMT", $year + 1900, $month + 1, $day, $hour, $minute);
        $worksheet->write($row + 1, 0, "Generated by RackMonkey $VERSION at $currentDate.", $footerFormat);

    }
    else
    {
        die "RM2XLS: Not a valid view for rack2xls.";
    }
};
if ($@)
{
    my $errMsg = $@;
    print $cgi->header;
    my $friendlyErrMsg = RackMonkey::Error::enlighten($errMsg);
    RackMonkey::Error::display($errMsg, $friendlyErrMsg);
}

sub formatSpreadsheet
{
    my $workbook = shift;

    # Add custom colors
    my $grey = $workbook->set_custom_color(40, '#282828');

    # Add and define default format
    my $format = $workbook->addformat(
        'font' => 'Verdana',
        'size' => 10
    );

    my $textwrap_format = $workbook->addformat();
    $textwrap_format->copy($format);
    $textwrap_format->set_text_wrap();

    $format->set_align('left');

    # Add and define url format
    my $url_format = $workbook->addformat();
    $url_format->set_font('Arial');
    $url_format->set_size(10);
    $url_format->set_align('left');
    $url_format->set_underline();

    # Add and define headers format
    my $headers_format = $workbook->addformat();
    $headers_format->set_font('Verdana');
    $headers_format->set_size(12);
    $headers_format->set_align('left');
    $headers_format->set_bg_color($grey);
    $headers_format->set_color('white');

    my $product_header = $workbook->addformat(
        'border'   => 5,
        'valign'   => 'vcenter',
        'align'    => 'left',
        'font'     => 'Arial',
        'size'     => 15,
        'bold'     => 3,
        'bg_color' => 'grey',
    );

    my $footerFormat = $workbook->addformat(
        'valign' => 'vcenter',
        'align'  => 'left',
        'font'   => 'Verdana',
        'size'   => 9,
        'bold'   => 3,
        'italic' => 1
    );

    return ($format, $headers_format, $footerFormat, $url_format);
}

# We can't have multiple formats in once cell so we have to strip formatting from notes
sub formatNotes
{
    my $note = shift;

    $note =~ s/\[(.*?)\|(.*?)\]/$2 ($1)/sg;    # format URLs
    $note =~ s/\*\*\*(.*?)\*\*\*/$1/sg;        # strip strong using ***
    $note =~ s/\*\*(.*?)\*\*/$1/sg;            # strip emphasis using **
    return $note;
}
